### PR TITLE
Femtoworld

### PR DESCRIPTION
**Pull request by zchochul**

### CODE REVIEW

**Comment by saganatt** [comment]
Can you remove redundant AliFemto and O2Femto directories?

**Comment by victor-gonzalez** [comment]
You  need to rebase your changes to the latest tag

**Review by victor-gonzalez — CHANGES_REQUESTED** [review]
Please, organize your code with the same directory layout as the `PWGCF/MultiparticleCorrelations` directory
`Core` for your internal classes definition
`DataModel` for your internal columns/tables definitions
`TableProducer` for your tasks which generate internal tables to be consumed by other task
`Tasks` for your analysis tasks

**Comment by saganatt** [comment]
@victor-gonzalez we copied the FemtoDream so we organized the code like them.

Will later all CF frameworks have the same structure? We will of course conform in this case :)

**Comment by lgraczykCern** [comment]
@victor-gonzalez this is still a part of our learning of O2. As @saganatt wrote, our code for the moment is a copy of the FemtoDream that we will start adjusting to our needs for Phi and D meson analysis, but first, we need to test what we have for the moment and also learn how to use Hyperloop.

**Comment by victor-gonzalez** [comment]
I understand it but it is better to start with the right structure from the beginning than correct it later

**Comment by victor-gonzalez** [comment]
@saganatt yes, all CF frameworks have to have the same structure
It is better to start with it from the beginning 
`MultiparticleCorrelations` also came with a flat structure as the one you intended and took them less than one hour to follow the right layout

**Comment by lgraczykCern** [comment]
OK @victor-gonzalez, it's fine; we will rearrange the structure to match what you propose. Indeed it is true it will make some order in the files and you're right, better do it from the beginning.

File: PWGCF/FemtoWorld/Core/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+# See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+# All rights not expressly granted are reserved.
+#
+# This software is distributed under the terms of the GNU General Public
+# License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+#
+# In applying this license CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
```
Comment (victor-gonzalez) [inline PWGCF/FemtoWorld/Core/CMakeLists.txt:10]: Have in mind that you don't necessarily need to only have headers files in your `Core` directory. 
You could also have `.cxx` files when needed and in that case you should instruct the compiler to create a library that you will link with your table producer tasks or with your task as needed
Check https://github.com/AliceO2Group/O2Physics/blob/master/PWGCF/Core/CMakeLists.txt for inspiration

File: PWGCF/FemtoWorld/Core/FemtoWorldContainer.h [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file FemtoWorldContainer.h
+/// \brief Definition of the FemtoWorldContainer
+/// \author Andi Mathis, TU München, andreas.mathis@ph.tum.de
+/// \author Valentina Mantovani Sarti, valentina.mantovani-sarti@tum.de
+
+#ifndef ANALYSIS_TASKS_PWGCF_FEMTOWORLD_FEMTOWORDLCONTAINER_H_
+#define ANALYSIS_TASKS_PWGCF_FEMTOWORLD_FEMTOWORDLCONTAINER_H_
```
Comment (victor-gonzalez) [inline PWGCF/FemtoWorld/Core/FemtoWorldContainer.h:18]: Probably it does make sense to get rid of the `ANALYSIS_TASKS_PWGCF_FEMTOWORLD_` head
What do you think?

File: PWGCF/FemtoWorld/Core/FemtoWorldDetaDphiStar.h [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file FemtoWorldDetaDphiStar.h
+/// \brief FemtoWorldDetaDphiStar - Checks particles for the close pair rejection.
+/// \author Laura Serksnyte, TU München, laura.serksnyte@tum.de
+
+#ifndef ANALYSIS_TASKS_PWGCF_O2FEMTO_O2FEMTODETADPHISTAR_H_
+#define ANALYSIS_TASKS_PWGCF_O2FEMTO_O2FEMTODETADPHISTAR_H_
```
Comment (victor-gonzalez) [inline PWGCF/FemtoWorld/Core/FemtoWorldDetaDphiStar.h:17]: The same as above and removing additionally the `O2`

File: PWGCF/FemtoWorld/Core/FemtoWorldEventHisto.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file FemtoWorldEventHisto.h
+/// \brief FemtoWorldEventHisto - Histogram class for tracks, V0s and cascades
+/// \author Andi Mathis, TU München, andreas.mathis@ph.tum.de
+
+#ifndef ANALYSIS_TASKS_PWGCF_FEMTOWORLD_INCLUDE_FEMTOWORLD_FEMTOEVENTHISTO_H_
+#define ANALYSIS_TASKS_PWGCF_FEMTOWORLD_INCLUDE_FEMTOWORLD_FEMTOEVENTHISTO_H_
```
Comment (victor-gonzalez) [inline PWGCF/FemtoWorld/Core/FemtoWorldEventHisto.h:17]: The same but in this case with `ANALYSIS_TASKS_PWGCF_FEMTOWORLD_INCLUDE_FEMTOWORLD_`

File: PWGCF/FemtoWorld/Core/FemtoWorldMath.h [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file FemtoWorldMath.h
+/// \brief Definition of the FemtoWorldMath Container for math calculations of quantities related to pairs
+/// \author Valentina Mantovani Sarti, TU München, valentina.mantovani-sarti@tum.de, Laura Serksnyte, TU München, laura.serksnyte@cern.ch
+
+#ifndef ANALYSIS_TASKS_PWGCF_O2FEMTO_O2FEMTOMATH_H_
+#define ANALYSIS_TASKS_PWGCF_O2FEMTO_O2FEMTOMATH_H_
```
Comment (victor-gonzalez) [inline PWGCF/FemtoWorld/Core/FemtoWorldMath.h:17]: Same as above but with `ANALYSIS_TASKS_PWGCF_O2FEMTO_O2`

File: PWGCF/FemtoWorld/Core/FemtoWorldPairCleaner.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file FemtoWorldPairCleaner.h
+/// \brief FemtoWorldPairCleaner - Makes sure only proper candidates are paired
+/// \author Andi Mathis, TU München, andreas.mathis@ph.tum.de, Laura Serksnyte <laura.serksnyte@cern.ch>, TU München
+
+#ifndef ANALYSIS_TASKS_PWGCF_FEMTOWORLD_INCLUDE_FEMTOWORLD_FEMTOWORLDPAIRCLEANER_H_
+#define ANALYSIS_TASKS_PWGCF_FEMTOWORLD_INCLUDE_FEMTOWORLD_FEMTOWORLDPAIRCLEANER_H_
```
Comment (victor-gonzalez) [inline PWGCF/FemtoWorld/Core/FemtoWorldPairCleaner.h:17]: The same as above but with `ANALYSIS_TASKS_PWGCF_FEMTOWORLD_INCLUDE_FEMTOWORLD_`

File: PWGCF/FemtoWorld/Core/FemtoWorldParticleHisto.h [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file FemtoWorldParticleHisto.h
+/// \brief FemtoWorldParticleHisto - Histogram class for tracks, V0s and cascades
+/// \author Andi Mathis, TU München, andreas.mathis@ph.tum.de
+
+#ifndef ANALYSIS_TASKS_PWGCF_FEMTOWORLD_INCLUDE_FEMTOWORLD_FEMTOPARTICLEHISTO_H_
+#define ANALYSIS_TASKS_PWGCF_FEMTOWORLD_INCLUDE_FEMTOWORLD_FEMTOPARTICLEHISTO_H_
```
Comment (victor-gonzalez) [inline PWGCF/FemtoWorld/Core/FemtoWorldParticleHisto.h:17]: The same as above but with `ANALYSIS_TASKS_PWGCF_FEMTOWORLD_INCLUDE_FEMTOWORLD_`

File: PWGCF/FemtoWorld/Core/FemtoWorldUtils.h [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file CFFilter.cxx
+/// \brief Utilities for the FemtoWorld framework
+///
+/// \author Luca Barioglio, TU München, luca.barioglio@cern.ch
+
+#ifndef ANALYSIS_TASKS_PWGCF_FEMTOWORLD_UTILS_H_
+#define ANALYSIS_TASKS_PWGCF_FEMTOWORLD_UTILS_H_
```
Comment (victor-gonzalez) [inline PWGCF/FemtoWorld/Core/FemtoWorldUtils.h:18]: The same as above but with `ANALYSIS_TASKS_PWGCF_`

File: PWGCF/FemtoWorld/Core/FemtoWorldUtils.h [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+/// function to retrieve the nominal mgnetic field in kG (0.1T) and convert it directly to T
+/// \param timestamp timestamp corresponding the the collision
+/// \param ccdb CCDB object that contains the configuration for the event
+/// \return magnetic field in Tesla
+
+float getMagneticFieldTesla(const uint64_t& timestamp, const Service<o2::ccdb::BasicCCDBManager>& ccdb)
+{
+  // TODO done only once (and not per run). Will be replaced by CCDBConfigurable
+  static o2::parameters::GRPObject* grpo = ccdb->getForTimeStamp<o2::parameters::GRPObject>("GLO/GRP/GRP", timestamp);
```
Comment (victor-gonzalez) [inline PWGCF/FemtoWorld/Core/FemtoWorldUtils.h:110]: Have in mind that for Run 3 the `CCDB` endpoint and the way of getting the magnetic field information has changed
Have a look at the implementation of PR#976 https://github.com/AliceO2Group/O2Physics/pull/976/files for inspiration
If you are not addressing Run 3 datasets for the time being you don't need to consider it. Just have it in mind

**Review by victor-gonzalez — COMMENTED** [review]
Please have a look at my comments
And thanks for changing the layout!!

**Review by saganatt — APPROVED** [review]
Hi @victor-gonzalez thank you very much for your comments!
We will indeed improve the ifndef guard clauses, they got a bit messy. And we will look at the CCDB later.
Since it doesn't affect the code performance now, I hope you doesn't mind merging this as-it, to be available on Hyperloop tomorrow? And tomorrow Zuzanna will fix the header files.

**Comment by victor-gonzalez** [comment]
It does make sense to me!

**Review by victor-gonzalez — APPROVED** [review]

**Pull request merged**